Wikipedia/Code/TalkPageTopicComposeViewController+TalkPageFormattingToolbar.swift

```diff
@@ -0,0 +1,118 @@
+import Foundation
+import CocoaLumberjackSwift
+import WMF
+
+extension TalkPageTopicComposeViewController: TalkPageFormattingToolbarViewDelegate {
+
+    func didSelectBold() {
+        bodyTextView.addOrRemoveStringFormattingCharacters(formattingString: "'''")
+    }
+
+    func didSelectItalics() {
+        bodyTextView.addOrRemoveStringFormattingCharacters(formattingString: "''")
+    }
+
+    func didSelectInsertLink() {
+        if let range =  bodyTextView.selectedTextRange {
+            let text = bodyTextView.text(in: range)
+            preselectedTextRange = range
+
+            var doesLinkExist = false
+
+            if let start = bodyTextView.position(from: range.start, offset: -2),
+               let end = bodyTextView.position(from: range.end, offset: 2),
+               let newSelectedRange = bodyTextView.textRange(from: start, to: end) {
+
+                if let newText = bodyTextView.text(in: newSelectedRange) {
+                    if newText.contains("[") || newText.contains("]") {
+                        doesLinkExist = true
+                    } else {
+                        doesLinkExist = false
+                    }
+                }
+            }
+
+            guard let link = Link(page: text, label: text, exists: doesLinkExist) else {
+                return
+            }
+
+            if link.exists {
+                guard let editLinkViewController = EditLinkViewController(link: link, siteURL: viewModel.siteURL, dataStore: MWKDataStore.shared()) else {
+                    return
+                }
+                editLinkViewController.delegate = self
+                let navigationController = WMFThemeableNavigationController(rootViewController: editLinkViewController, theme: self.theme)
+                navigationController.isNavigationBarHidden = true
+                present(navigationController, animated: true)
+            } else {
+                let insertLinkViewController = InsertLinkViewController(link: link, siteURL: viewModel.siteURL, dataStore: MWKDataStore.shared())
+                insertLinkViewController.delegate = self
+                let navigationController = WMFThemeableNavigationController(rootViewController: insertLinkViewController, theme: self.theme)
+                present(navigationController, animated: true)
+            }
+        }
+    }
+
+}
+
+extension TalkPageTopicComposeViewController: InsertLinkViewControllerDelegate {
+    func insertLinkViewController(_ insertLinkViewController: InsertLinkViewController, didTapCloseButton button: UIBarButtonItem) {
+        dismiss(animated: true)
+    }
+
+    func insertLinkViewController(_ insertLinkViewController: InsertLinkViewController, didInsertLinkFor page: String, withLabel label: String?) {
+        insertLink(page: page)
+        dismiss(animated: true)
+    }
+
+    func insertLink(page: String) {
+        let content = "[[\(page)]]"
```

Comment:
What do you think about something like this here (and in the other spot)?

```
        var content: String = "[[\(page)]]"
        
        if let selectedText = bodyTextView.text(in: preselectedTextRange) {
            if page != selectedText {
                content = "[[\(page)|\(selectedText)]]"
            }
        }
```

This way users can get some of the piping wikitext to paste without going through the EditLinkViewController. (This also seems to be how our section editor works). Here's what it does:

![out](https://user-images.githubusercontent.com/3620196/206037477-836e7028-0d99-463c-ab63-6b3fe9df4c5a.gif)